clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java

```diff
@@ -164,105 +94,80 @@ public OffsetFetchResponse(Errors error, Map<TopicPartition, PartitionData> resp
+                                       .setCommittedLeaderEpoch(partitionData.leaderEpoch.orElse(0))
+                                       .setMetadata(partitionData.metadata)
+            );
+            offsetFetchResponseTopicMap.put(topicName, topic);
         }
 
-        // for version 2 and later use the top-level error code (in ERROR_CODE_KEY_NAME) from the response.
-        // for older versions there is no top-level error in the response and all errors are partition errors,
-        // so if there is a group or coordinator error at the partition level use that as the top-level error.
-        // this way clients can depend on the top-level error regardless of the offset fetch version.
-        this.error = struct.hasField(ERROR_CODE) ? Errors.forCode(struct.get(ERROR_CODE)) : topLevelError;
+        this.data = new OffsetFetchResponseData()
+            .setTopics(new ArrayList<>(offsetFetchResponseTopicMap.values()))
+            .setErrorCode(error.code())
+            .setThrottleTimeMs(throttleTimeMs);
     }
 
-    public void maybeThrowFirstPartitionError() {
-        Collection<PartitionData> partitionsData = this.responseData.values();
-        for (PartitionData data : partitionsData) {
-            if (data.hasError())
-                throw data.error.exception();
-        }
+    public OffsetFetchResponse(Struct struct, short version) {
+        this.data = new OffsetFetchResponseData(struct, version);
     }
 
     @Override
     public int throttleTimeMs() {
-        return throttleTimeMs;
+        return data.throttleTimeMs();
     }
 
     public boolean hasError() {
-        return this.error != Errors.NONE;
+        return this.data.errorCode() != Errors.NONE.code();
     }
 
     public Errors error() {
-        return this.error;
+        return Errors.forCode(data.errorCode());
     }
 
     @Override
     public Map<Errors, Integer> errorCounts() {
-        return errorCounts(error);
+        Map<TopicPartition, Errors> errorMap = new HashMap<>();
+        for (OffsetFetchResponseTopic topic : data.topics()) {
+            for (OffsetFetchResponsePartition partition : topic.partitions()) {
+                errorMap.put(new TopicPartition(topic.name(), partition.partitionIndex()),
+                             Errors.forCode(partition.errorCode()));
+            }
+        }
+        return errorCounts(errorMap);
     }
 
     public Map<TopicPartition, PartitionData> responseData() {
+        Map<TopicPartition, PartitionData> responseData = new HashMap<>();
+        for (OffsetFetchResponseTopic topic : data.topics()) {
+            for (OffsetFetchResponsePartition partition : topic.partitions()) {
+                responseData.put(new TopicPartition(topic.name(), partition.partitionIndex()),
+                                 new PartitionData(partition.committedOffset(),
+                                                   Optional.of(partition.committedLeaderEpoch()),
```

Comment:
Hmm.. `partition.committedLeaderEpoch()` is an int, so this will never be `Optional.empty()`. Shouldn't we be using `RequestUtils.getLeaderEpoch`?